Refactor remaining Project and Build files

Nirmal4G opened the pull request:
### Info

Extracted from CommunityToolkit/WindowsCommunityToolkit#4234 as the Common Code projects were separated in this new repository.


### Changes

- Simplify project files
  - Remove unnecessary lines
  - Add and Update comments
  - Rearrange and Order code blocks

- Include build root folder
- Rename NuGet Icon to `Icon.png`
  - This is no longer used as a public reference point for NuGet icon URL. Hence, it's safe to change.
- Normalize casing for `ReadMe.md`
  - To match the file already in [CommunityToolkit/WindowsCommunityToolkit@`main`/ReadMe.md](/CommunityToolkit/WindowsCommunityToolkit/blob/main/ReadMe.md)
- Rename solution file to `CommunityToolkit.sln`
  - Spaces are a main issue when doing automation (_like using `*.sln` in build scripts_).
  - Also, `dotnet` seems implied and doesn't stand-out in the file list because of the lower casing.


### Notes

Please suggest alternate names for `dotnet Community Toolkit.sln`. We have already established that we can't use `.NET`.
I tried using `NET`, `DotNET`, `dotNET` and `Core` as prefix. Except for `Core`, others neither look good nor stand-out in the file list.

- **Rebase** merge please.
- When merging, please update the commit title to PR title instead of the default `Merge pull request #xxxx from repo/branch`, and commit message to either PR message or messages of individual commits. The `auto-merge` option does this by default.

Nirmal4G commented:
### Project File structuring

- SOF (_Start Of File_) Imports,
- Core properties, Package properties, build properties,
- Build items, Resource items, Content items, Misc. items,
- In-place Imports, Choose (_elements within follows outer sort order_)
- `ProjectReference` items, `PackageReference` items,
- Targets and EOF (_End Of File_) Imports.

Where there's a condition by target properties, we should group them together under `Choose`. For multiple target values, we should sort them by the order in which they are defined. And the order in which they should be defined is either ascending or descending in terms of compatibility layering.

Nirmal4G commented:
@Sergio0694 PR closed by 237 by mistake. Can you reopen it?

Sergio0694 commented:
Huh, weird, yeah no problem! 😄

Sergio0694 commented on tests/CommunityToolkit.Mvvm.ExternalAssembly/CommunityToolkit.Mvvm.ExternalAssembly.csproj:
```diff
@@ -4,8 +4,11 @@
-  <!-- Manually apply the properties of test projects to this assembly as well,
-       which is neither a test project not one of the libraries to pack for NuGet. -->
+  <!--
+      MVVM unit test uses this project to test it as a `ProjectReference`.
+      Since, we don't need to pack or publish the project, disable those
+      targets and ignore validation warnings given in those contexts.
```
Drop the ` as they don't do anything in XML.
Also, please keep comments impersonal.
```suggestion
      The MVVM Toolkit unit tests use this project to test it as a ProjectReference.
      Since this project is neither packed nor published, those targets can be
      disabled, and validation warnings can be ignored in those contexts.
```

Nirmal4G replied:
The backtick I use in comments to represent code references, even though, the editors can't do anything now but it's future proof. Also, one of the extensions I use actually highlight this separately.

Sergio0694 replied:
Please just don't use ` and remove them from the places where you've added that 🙂
It's not about future proofing, it's just not part of XML, so it shouldn't be there.

Nirmal4G replied:
XML text doesn't have a specific restriction on symbols but if you don't like backtick, I'll replace it with standard single quote as I did in other places. This was just an experiment to get feedback.

Sergio0694 replied:
I would appreciate if you could just follow the review feedbacks here, as it'd save a lot of back and forth. I've asked you multiple times to just leave those identifiers as they were. Can you please just do that? Don't add anything.

Nirmal4G replied:
Sure. I was only trying to improve the readability.

Sergio0694 commented on tests/CommunityToolkit.Mvvm.ExternalAssembly/CommunityToolkit.Mvvm.ExternalAssembly.csproj:
```diff
@@ -14,7 +17,8 @@
   <ItemGroup>
     <ProjectReference Include="..\..\CommunityToolkit.Mvvm\CommunityToolkit.Mvvm.csproj" />
-    <ProjectReference Include="..\..\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" PrivateAssets="contentfiles;build" />
+    <ProjectReference Include="..\..\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.csproj"
```
These (and same for the other references in other files) should just be a single line.
VS keeps updating them anyway all the time, so better to just get them to one line from the start.

Nirmal4G replied:
VS shouldn't do that, at least for SDK-style projects. Let me check this.

Sergio0694 reviewed (changes requested):
Left some nits 🙂

Nirmal4G commented on CommunityToolkit.Diagnostics/CommunityToolkit.Diagnostics.csproj:
```diff
@@ -34,7 +36,10 @@
 
-    <!-- Enable trimming support on .NET 6 -->
+    <!--
+        NETSTANDARD2_1_OR_GREATER: includes both .NET Standard 2.1 and .NET 6.
```
Sure, I'll simplify the comments.

Nirmal4G replied:
I don't think the message mean same as everywhere. I'll try to simplify the comments.

Nirmal4G replied:
Updated the comments!

Nirmal4G replied:
We don't need to specify the pre-processor symbol for the .NET Standard as it is implicitly included! I've also updated the comment to be clearer.

Nirmal4G commented on CommunityToolkit.HighPerformance/CommunityToolkit.HighPerformance.csproj:
```diff
@@ -39,33 +39,42 @@
 
-      <!-- NETSTANDARD2_1_OR_GREATER: includes both .NET Standard 2.1, .NET Core 3.1 and .NET 6.
-           Additionally, also enable trimming support on .NET 6. -->
+      <!-- NETSTANDARD2_1_OR_GREATER: includes both .NET Standard 2.1, .NET Core 3.1 and .NET 6 -->
```
There's difference here but I'll try to simplify them.

Nirmal4G commented on CommunityToolkit.Mvvm.SourceGenerators/Attributes/NotNullWhenAttribute.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
```
EditorConfig ensures that. These files are leftovers from the original repo without the formatting patch properly applied.

Nirmal4G commented on CommunityToolkit.Mvvm.SourceGenerators/CommunityToolkit.Mvvm.SourceGenerators.csproj:
```diff
@@ -1,43 +1,29 @@
 
+  <!--
+      These files enables tracking Analyzer releases for the rules that ship from this project.
+      See <https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md>
```
Having dot at the end, some editors might treat it as part of the URL. Most support angle bracket notation even if those editors are not markdown complaint.

Sergio0694 replied:
I see. Still don't like the `<>` though as they're not standard XML. Can we just do:
```xml
See https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md for more info.
```
No `<>`, and not need for `.` after the URL either.

Nirmal4G replied:
Sure, why not! I was just looking forward.

Nirmal4G commented on CommunityToolkit.Mvvm/CommunityToolkit.Mvvm.csproj:
```diff
@@ -34,30 +48,21 @@
 
+  <!--
+      Pack the source generator to the right package folder.
+      TODO: Find a way to get the analyzer project's build output path,
```
I already did. It's the next PR I opened. We can leave this as is. It should provide some context.

Sergio0694 replied:
It's in the next PR you opened, but we haven't even reviewed nor approved it yet, so this PR should not make assumptions about what other code may or may not be added. The current comment is fine on its own, without the TODO.

Nirmal4G replied:
No, I'm not making assumptions. Just through TODO, I'm emphasizing that there should be a better way. I'll update the comment to be clearer and mention the current implementation is HACK! Would that be alright with you?

Nirmal4G replied:
No, I'm not making assumptions. Just through TODO, I'm emphasizing that there should be a better way. I'll update the comment to be clearer and mention the current implementation is a HACK!

Nirmal4G replied:
No, I'm not making assumptions. Just through TODO, I'm emphasizing that there should be a better way. I've updated the comment to be clearer and mentioned that the current implementation is a HACK!

Nirmal4G commented on CommunityToolkit.Mvvm/CommunityToolkit.Mvvm.csproj:
```diff
@@ -34,30 +48,21 @@
           PackagePath="analyzers\dotnet\roslyn4.0\cs"
-          Pack="true"
-          Visible="false" />
+          Pack="true" Visible="false" />
```
It was just a side effect of making the patch tree diff-able and blame-able. The next PR actually removes all this. So, this doesn't matter here.

Nirmal4G commented on CommunityToolkit.Mvvm/CommunityToolkit.Mvvm.targets:
```diff
@@ -13,20 +13,20 @@
       <CurrentCompilerVersion>@(CurrentCompilerAssemblyIdentity->'%(Version)')</CurrentCompilerVersion>
-
-      <!-- The CurrentCompilerVersionIsNotNewEnough property can now be defined based on the Roslyn assembly version -->
+      <!-- The 'CurrentCompilerVersionIsNotNewEnough' property can now be defined based on the Roslyn assembly version -->
```
Sure, will do.

Nirmal4G replied:
Added the new line back!

Nirmal4G replied:
The new lines are back!

Nirmal4G commented on Directory.Build.props:
```diff
@@ -36,7 +36,7 @@
         <EmbedUntrackedSources>true</EmbedUntrackedSources>
         <!-- Include PDB in the built .nupkg -->
-        <AllowedOutputExtensionsInPackageBuildOutputFolder>$(AllowedOutputExtensionsInPackageBuildOutputFolder);.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>
+        <AllowedOutputExtensionsInPackageBuildOutputFolder>.pdb</AllowedOutputExtensionsInPackageBuildOutputFolder>
```
There's no previous value. It's redundant to have a previous value here when it's always empty!

Nirmal4G commented on build/Community.Toolkit.Common.targets:
```diff
@@ -12,9 +12,9 @@
-    <None Include="$(BuildToolsDirectory)nuget.png" Pack="true" PackagePath="\Icon.png" Visible="False" />
-    <None Include="$(RepositoryDirectory)License.md" Pack="true" PackagePath="\" Visible="False" />
-    <None Include="$(RepositoryDirectory)ThirdPartyNotices.txt" Pack="true" PackagePath="\" Visible="False" />
+    <None Pack="true" PackagePath="\" Visible="False" Include="$(BuildToolsDirectory)Icon.png" />
```
I want, with all of my being, to start with `Include` but doesn't **this look nice** and it's not like we're going to update this anytime soon!

Sergio0694 replied:
No it doesn't, please just leave `Include` first 😆
It's also alphabetically in order, I don't understand why you would want to put that last at all.

Nirmal4G replied:
I actually forgot why I changed this, since it was a long time ago. But thanks to [this comment](#discussion_r888842771), I found that I missed a commit during rebase from the split from the **WindowsCommunityToolkit** repo. It is now included here.

Nirmal4G commented:
Fixed the conflicts introduced by string-name PR!

Nirmal4G commented:
Fixed the conflicts introduced by strong-name PR!

Sergio0694 commented on CommunityToolkit.Diagnostics/CommunityToolkit.Diagnostics.csproj:
```diff
@@ -44,6 +44,12 @@
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
+  </ItemGroup>
+
+  <!-- Text Templates to generate sources -->
```
```suggestion
  <!-- Text templates to generate sources -->
```

Nirmal4G replied:
Since the extension is `tt`, I capitalised the second word to emphasize that.

Sergio0694 replied:
I'm aware of that. As I mentioned, please, let's try to avoid back and forth for every single review comment.
This whole PR is already taking way too much time for something that I don't even think is that worth it to be honest.

Sergio0694 commented on CommunityToolkit.Mvvm/CommunityToolkit.Mvvm.csproj:
```diff
@@ -20,44 +20,55 @@
+
+  <!--
+      Include a custom targets file to check the included source generator.
+      Including it in .NET 6+ targets is not needed as it guarantees Roslyn 4.x!
```
```suggestion
      Including it in .NET 6+ targets is not needed as it guarantees Roslyn 4.x.
```

Sergio0694 commented on tests/CommunityToolkit.Mvvm.DisableINotifyPropertyChanging.UnitTests/CommunityToolkit.Mvvm.DisableINotifyPropertyChanging.UnitTests.csproj:
```diff
@@ -5,14 +5,14 @@
-    <PackageReference Include="MSTest.TestFramework" Version="2.2.8" />
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
+    <ProjectReference Include="..\..\CommunityToolkit.Mvvm\CommunityToolkit.Mvvm.csproj" />
+    <ProjectReference Include="..\..\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" PrivateAssets="contentfiles;build" />
```
Some analyzer references are inconsistent. Here it's all in one line, and in other places they're split differently. We should just pick one style and always use that one.

Nirmal4G replied:
This might've been left out during rebase. I'll fix that.

Nirmal4G commented:
Fixed the conflicts introduced by PRs 290, 292 and a package update! Also included a missed commit during rebase. see [this comment on why](#discussion_r892601766). It should also explain why (_To keep the diff simple between each commit_) some lines have been rearranged.

_Let me know if you want to separate the last commit into a new PR since the changes are too broad._

Nirmal4G commented:
The compare option beside the force-pushed status is broken. It doesn't show the diff properly. Use Files changed tab.

Nirmal4G commented:
The compare option beside the force-pushed status is broken. It doesn't show the diff properly.

Use this compare for changes since the last review: [`61c5ea7..cf577f0`](/CommunityToolkit/dotnet/compare/61c5ea7..cf577f0)

Sergio0694 commented:
@Nirmal4G This is just way too much for a single PR. Also, you just cannot keep adding new stuff after opening a PR. How is anyone supposed to leave a review if you keep changing more and more stuff after people have already started looking at the changes. This just forces people to review everything from scratch all over again, and it's really time consuming. Not to mention, the PR title says "refactoring", but clearly it's not doing just that. This PR cannot be reviewed or merged this way.

If you want to introduce some functional/build change, please open an issue first so we can discuss things, and _then_ open a PR _just for that change_. You can't just keep adding more and more stuff to this PR, it causes too much churn and ovehead for us.

Nirmal4G commented:
@Sergio0694 I have edited this patch for easy to review. If you want, you can review commit by commit or review only [**changes since last review**](/Nirmal4G/appkit-dotnet/compare/archive/refactor..hotfix/refactor). The `Compare` button beside the `force-push` event listed here doesn't work properly.

Changes since last review: [Nirmal4G/appkit-dotnet@`archive/refactor..hotfix/refactor`](https://github.com/Nirmal4G/appkit-dotnet/compare/archive/refactor..hotfix/refactor)

Sergio0694 commented:
This PR still doing way more than the title says.
It's not just refactoring stuff, it's also making a whole lot of functional changes.

I should also repeat: I consider build changes to be very high-risk.
I'm not seeing a benefit of changing so much stuff here. If there is a specific issue you think should be solved, then you should first open an issue with a proper explanation so we can discuss things, and only _then_ open a PR once we all agree.

Nirmal4G commented:
Is there any particular change you want to discuss? The only build change is the last commit that introduces a new Item type for packing common files regardless of pack content option! That's the only change but it's non-breaking.

Sergio0694 commented:
This PR includes a whole bunch of MSBuild changes and new targets.
You can't just slip those kind of changes into a PR just saying it's "refactoring" stuff.

Again, if the PR says it's doing refactoring, there should be **no functional changes**.

Nirmal4G commented:
**Except last commit, there are no functional changes**. Look carefully, I just simplified, re-arranged and extracted common MSBuild logic to a new file. That's all. Even the last commit is not functionally different but I do get that it's new. The last might be related to the Dynamic Pack PR I have; I can remove it from here if you're okay with that.

Sergio0694 commented:
- Moving the T4 stuff into a new .target should be its own PR.
- The whole last commit should be its own PR.
- Changing the lang version and the compiler strictness mode are functional changes.

This whole PR is like 4 different PRs all combined. And as I said, especially for the whole last commit, you should first open an issue to illustrate the issue, and _then_ a PR if that's approved. It's not even clear to me what actual problem is the last commit trying to solve in the first place. This PR, as I said multiple times already, is just doing way more than it states.

I should also repeat, doing this literally just makes it more difficult to also review things, meaning it makes us waste time, and it makes you waste time as well for all the back and forth. It's just not the best way to contribute to the project. I appreciate your enthusiasm for wanting to help here, but this is really not the best way to go about it.

Nirmal4G commented:
> Moving the T4 stuff into a new .target should be its own PR.

That is called refactoring in MSBuild world! 😎

> The whole last commit should be its own PR.

I agree. It's related to the Dynamic Pack PR I already have. Can I leave it there? since that PR requires changes from this commit. So, the two must be reviewed and accepted together.

> Changing the lang version and the compiler strictness mode are functional changes.

The binaries are the same. That's why I included this change in this PR. I can move those into the next PR or a new one.

Compiler strictness just enforces at language-level. But this Behavior changes only if we're using experimental and non-standard features. Lang version just sets it to latest, since we're always upgrading and not downgrading out .NET SDK versions, we can use the latest language features if and when it becomes available. Use of strictness also ensures that we use watered down, restricted feature set so if and when Roslyn breaks something (_I hope never_) we won't get burned because of that!

Nirmal4G commented:
> I should also repeat, doing this literally just makes it more difficult to also review things, meaning it makes us waste time, and it makes you waste time as well for all the back and forth. It's just not the best way to contribute to the project. I appreciate your enthusiasm for wanting to help here, but this is really not the best way to go about it.

**I'm so sorry about this.** I'm new to the whole contributing thing. You can infer most of how I do source control from my commits. Creating a whole new PR is like applying for a passport or a visa (😢), it's too much for every single change. That's why I make related changes in a single PR and sometimes they become big. I'm still learning what constitutes a lean and mean PR for a particular project. Earlier you put different style refactoring all in one PR, even though it breaks binary compatibility.

The kinds of changes I make here and to other projects out there, no one really does them because it's boring, repetitive and as you've said causes too much churn but we have to do it at least once in a while to make the project good for first-time learners such as myself!

Nirmal4G commented:
@Sergio0694 I have sent you a DM on Discord in regards to the updated patch tree here.

For only changes since the last review: https://github.com/CommunityToolkit/dotnet/compare/a3b2649..322bc61

Nirmal4G commented:
@Sergio0694 Sorry for the notification but is there anything blocking this PR from getting merged?

Nirmal4G commented:
@Sergio0694 Updated in response to PR #428 which broke this!

Sergio0694 commented:
Just a heads up, there's major changes I'm doing in #436, so this PR will need updates after that as well.

michael-hawker commented:
Still going to suggest that we close this PR, as it should be broken out into smaller PRs as mentioned above.

@Nirmal4G in general doing things like renaming a directory from `build` to `eng` has little purpose or value to the project and just creates noise in these PRs. It's something that should be discussed in an issue first, folks can agree on a timing and value of the proposal, and then it can be coordinated with others working on the repo for a major folder restructure between releases.

Similarly, unless we can automate structuring of the project file XML with a tool like XAML Styler does for XAML files in the WCT repo, then manually moving Item/PropertyGroups around is just a temporary thing that also just creates more noise that's hard to review. Especially when mixed in with other changes in part of a larger PR.

Diving into making changes can be fun and exciting, but if you want to ensure your contributions are having a meaningful impact and will be merged; it's best to start a conversation with an issue or discussion first to lay out exactly what you plan to do, how it benefits the project, and what's involved with the change. This should be scoped to something that's a single 'block' and not a laundry list of lots of items. Each idea should be its own issue.

Or ask to pick up an already open bug or issue on the repo that someone's not working on or reach out on Discord to Sergio or myself and ask if there's something on the top-of-our-minds build related that we're struggling with that you could assist with. Really happy to have your passion directed at our projects, but we just want to ensure we're on the same page and can move forward together! Thanks!

Nirmal4G commented:
@michael-hawker This is still in draft! Also this is on top of existing PRs. Changes that are mutually exclusive are separated but when they are cumulative, how they can be separated?

Code changes are one thing but refactorings  are if not always cumulative and to produce clean diff and blame, they need to be arranged properly by theme of change or purpose of change.

I'm still a novice but if may ask, how would you separate this patch without causing conflicts that break build!?

Nirmal4G commented:
@michael-hawker There's a discussion open for these PRs. It's part of a larger refactoring with goal of introducing a common build infra for Community Toolkit repos (_or atleast for my forks if you're not willing to move forward with the idea!_).

wzxprince commented:
I am Chinese！！！I like it！
